encoder/vaapiencoder_vp9.cpp

```diff
@@ -191,23 +192,54 @@ bool VaapiEncoderVP9::fill(VAEncPictureParameterBufferVP9* picParam,
             picParam->reference_frames[i] = VA_INVALID_SURFACE;
     }
     else {
-        picParam->refresh_frame_flags = 0x01; // refresh last frame
+
         picParam->pic_flags.bits.frame_type = kInterFrame;
 
         ReferenceQueue::const_iterator it = m_reference.begin();
 
-        for (uint32_t i = 0; it != m_reference.end(); ++it, i++)
+        for (uint32_t i = 0; it != m_reference.end(); ++it, i++) {
             picParam->reference_frames[i] = (*it)->getID();
+            DEBUG("reference frame[%d] 0x%x", i, (*it)->getID());
+        }
 
+        picParam->pic_flags.bits.frame_context_idx = 0;
         // last/golden/alt is used as reference frame. L0 forward
         picParam->ref_flags.bits.ref_frame_ctrl_l0 = 0x7;
 
         // golden and alt are last KeyFrame
         // last is last decoded frame
-        picParam->ref_flags.bits.ref_last_idx = 0;
-        picParam->ref_flags.bits.ref_gf_idx = 1;
-        picParam->ref_flags.bits.ref_arf_idx = 2;
-        picParam->pic_flags.bits.frame_context_idx = 0;
+
+        if (getReferenceMode()) {
+            // intel driver updates a new slot with every new frame and keeps
+            // the reference frames in a circular buffer, the buffer is defined
+            // with 8 slots but it is up to the application to use as many as
+            // desired. This scheme implements the use of 3 slots for
+            // last/gold/alt refs
+
+            m_currentReferenceIndex = (m_currentReferenceIndex + 1) % 3;
+
+            // set to refresh next slot with current reconstructed surface
+            picParam->refresh_frame_flags = 1 << m_currentReferenceIndex;
+            // assign the references on only 3 slots
+            picParam->ref_flags.bits.ref_arf_idx = m_currentReferenceIndex;
+            picParam->ref_flags.bits.ref_gf_idx = m_currentReferenceIndex + 1;
+            picParam->ref_flags.bits.ref_last_idx = m_currentReferenceIndex - 1;
+
+            if (m_currentReferenceIndex == 0) {
+                picParam->ref_flags.bits.ref_last_idx
+                    = m_currentReferenceIndex + 2;
```

Comment:
That is where the referenceIndex has to wrap around.  Same below, they are needed for this logic to work.